<!--
---
title: "cell-classifier"
author: "aopisco"
date: "4/16/2018"
output: html_document
---
-->

# Unsupervised methods for combining independent single-cell RNA sequencing experiments

### Pre-required installation
We start by installing the packages we need to build the cell classifiers from bioconductor:
```{r}
source("https://bioconductor.org/biocLite.R")
biocLite("Seurat")
biocLite("scmap")
```

and then load the required libraries

```{r}
library(here)
library(tidyverse)
library(stringr)

library(Seurat)
library(viridis)
library(caret)
library(mlbench)
library(e1071)

library(SingleCellExperiment)
library(scmap)
```

As we briefly mentioned in the previous post, instead of manually and arbitrarily deciding which cell ontologies we should be comparing and computing the overlaps at the level of differentially expressed genes, we want to have a method to map on the fly new coming experimental data to existing reference datasets.

Given the broad collection of organs in Tabula muris it is natural to think of it as a possible reference dataset. Here we will exemplify how we can build the reference classifier using the built-in functions of ```Seurat``` and the recently published ```scmap``` method.

We can get the Tabula muris data from [here](https://github.com/czbiohub/tabula-muris-vignettes/tree/experiment-comparisons/data):
```{r}
tm.droplet.matrix = readRDS(here("data-tabula-muris", "TM_droplet_mat.rds"))
tm.droplet.metadata = read_csv(here("data-tabula-muris", "TM_droplet_metadata.csv"))

tm.facs.matrix = readRDS(here("data-tabula-muris", "TM_facs_mat.rds"))
tm.facs.metadata = read_csv(here("data-tabula-muris", "TM_facs_metadata.csv"))
```
<!--
# 
# microwell.matrix = readRDS(here("data-microwell-seq", "MCA_merged_mat.rds"))
# microwell.metadata = read_csv(here("data-microwell-seq", "MCA_All-batch-removed-assignments.csv"))

```
-->

For speed reasons we will be demonstrating the methods for a single tissue. We can choose from the list of tissues available:
```{r}
tiss.name.tm.droplet = unique(tm.droplet.metadata$tissue)
tiss.name.tm.facs = unique(tm.facs.metadata$tissue)
```
<!--
# tiss.name.microwellseq = unique(microwell.metadata$Tissue)
```
-->

Let us use ```Bladder`` as the organ of interest:
```{r}
droplet.tiss.interest = "Bladder"
facs.tiss.interest = "Bladder"

if (droplet.tiss.interest %in% tiss.name.tm.droplet){
  tm.droplet.tiss <- tm.droplet.matrix[,tm.droplet.metadata$tissue==droplet.tiss.interest] 
}

if(facs.tiss.interest %in% tiss.name.tm.facs){
  tm.facs.tiss <- tm.facs.matrix[,tm.facs.metadata$tissue==facs.tiss.interest]
}

```
<!--
#microwell.bladder <- microwell.matrix[,microwell.metadata$Tissue=="Bladder"]

```
-->

## Seurat built-in functions for cell classification

We start by creating the Seurat objects for our data:
```{r}
tm.droplet.metadata.tiss <- tm.droplet.metadata[tm.droplet.metadata$tissue == droplet.tiss.interest,]
tm.droplet.tiss.ann <- as.data.frame(tm.droplet.metadata.tiss, row.names = tm.droplet.metadata.tiss$cell)
rownames(tm.droplet.tiss.ann) <- tm.droplet.tiss.ann[,1]
tissDROPLET <- CreateSeuratObject(raw.data = tm.droplet.tiss, meta.data = tm.droplet.tiss.ann)
```

```{r}
tm.facs.metadata.tiss <- tm.facs.metadata[tm.facs.metadata$tissue == facs.tiss.interest,]
tm.facs.tiss.ann <- as.data.frame(tm.facs.metadata.tiss, row.names = tm.facs.metadata.tiss$cell)
rownames(tm.facs.tiss.ann) <- tm.facs.tiss.ann[,1]
tissFACS <- CreateSeuratObject(raw.data = tm.facs.tiss, meta.data = tm.facs.tiss.ann)
```

Next we use the ```BuildRFClassifier``` function to build a random forest classifier for each experimental method. This step can take a long time if the number of classes we want is big, but for the current example it should only take few minutes. Investing the time at this point is practical because we can later re-use the same classifier with several sets of data:

<!--
#tissDROPLET@meta.data$cell_ontology_class <- replace(tissDROPLET@meta.data$cell_ontology_class, is.na(tissDROPLET@meta.data$cell_ontology_class), "notKnown") 
-->

```{r}
tm.droplet.tiss.train <- tm.droplet.tiss[,!is.na(tm.droplet.tiss.ann$cell_ontology_class)] 
tm.droplet.tiss.train.metadata <- tm.droplet.tiss.ann[!is.na(tm.droplet.tiss.ann$cell_ontology_class),]

train.tissDROPLET <- CreateSeuratObject(raw.data = tm.droplet.tiss.train, meta.data = tm.droplet.tiss.train.metadata)

DROPLET_classes = as_factor(tissDROPLET@meta.data$cell_ontology_class)
classifierDROPLET <- BuildRFClassifier(train.tissDROPLET, training.classes = train.tissDROPLET@meta.data$cell_ontology_class)
```


```{r}
tm.facs.tiss.train <- tm.facs.tiss[,!is.na(tm.facs.tiss.ann$cell_ontology_class)] 
tm.facs.tiss.train.metadata <- tm.facs.tiss.ann[!is.na(tm.facs.tiss.ann$cell_ontology_class),]

train.tissFACS <- CreateSeuratObject(raw.data = tm.facs.tiss.train, meta.data = tm.facs.tiss.train.metadata)

FACS_classes = as_factor(tissFACS@meta.data$cell_ontology_class)
classifierFACS <- BuildRFClassifier(train.tissFACS, training.classes = train.tissFACS@meta.data$cell_ontology_class)
```


Once we have the classifiers we can use them to either classify data from the dataset we build them on or new data.
We start by verifing our classifiers:
```{r}

train.tissFACS <- SubsetData(object = tissFACS, cells.use = tissFACS@cell.names[seq(11,length(tissFACS@cell.names),100)])

predicted.classes.tissFACS <- ClassifyCells(
  object = train.tissFACS,
  classifierFACS,
  training.genes = NULL,
  training.classes = NULL,
  new.data = tissFACS@data
)

confusionMatrix(data = predicted.classes.tissFACS, reference = FACS_classes)
```

```{r}


train.tissDROPLET <- SubsetData(object = tissDROPLET, cells.use = tissDROPLET@cell.names[seq(11,length(tissDROPLET@cell.names),100)])

predicted.classes.tissDROPLET <- ClassifyCells(
  object = train.tissDROPLET,
  classifierDROPLET,
  training.genes = NULL,
  training.classes = NULL,
  new.data = tissDROPLET@data
)

confusionMatrix(data = predicted.classes.tissDROPLET, reference = DROPLET_classes)

```

As expecte dthe accuracy here is 1 because we are classifying exactly the same data for which we built the object against to.
More insteresting is to classify the data from, e.g. droplet method using the classifier built agains the FACS-based dataset:

```{r}
predicted.classes.tissDROPLET_usingFACS <- ClassifyCells(
  object = train.tissFACS,
  classifierFACS,
  training.genes = NULL,
  training.classes = NULL,
  new.data = tissDROPLET@data
)

u = union(predicted.classes.tissDROPLET_usingFACS, DROPLET_classes)
confusionMatrix(data = factor(predicted.classes.tissDROPLET_usingFACS, u), reference = factor(DROPLET_classes, u))

```

This example illustrates one likely common problem that we have to deal with. The FACS-based dataset has only tow levels, that is, only 2 cell ontologies were annotated, ```bladder cell``` and ```bladder urothelial cell``. However, the droplet-based dataset was annotated for 4 cell ontologies. In this situation, the random forest based classifier does not have the knowledge of possibly existing more labels and so tries to guess where best those cells would fit in out of the available possibilities.
In reverse, using the droplet classifier on the FACS data enables us to question our annotations. This is the opposite situation, here the classifier has more boxes where it can fit cells in than the original number of boxes in the dataset.

```{r}
predicted.classes.tissFACS_usingDROPLET <- ClassifyCells(
  object = train.tissDROPLET,
  classifierDROPLET,
  training.genes = NULL,
  training.classes = NULL,
  new.data = tissFACS@data
)

u = union(predicted.classes.tissFACS_usingDROPLET, FACS_classes)
confusionMatrix(data = factor(predicted.classes.tissFACS_usingDROPLET, u), reference = factor(FACS_classes, u))

```

Interestingly enough some cells are classified into an ontology different from the reference one.
Of note, the union trick is necessary to account for the fact that the number of levels in the data and in the classifier might not match.

## scmap
As we briefly mentioned above, using the random forest classifiers will *always* allocate a cell to one of the available levels but sometimes that might not be the answer we are looking for. It is well-known that cells existing in a continuum of states, and ideally we want to have a method that captures this by either classifying the cell in question as an established cell type or instead providing us the information that such cell cannot be classified and also the closest possible classifications.
[scmap](https://www-nature-com.ucsf.idm.oclc.org/articles/nmeth.4644) can map individual cells from a query sample to cell types in the reference (scmap-cluster) or to individual cells in a reference (scmap-cell).

### scmap input
If you already have a SingleCellExperiment object, skip this but if you have a matrix or a data frame containing expression data then you first need to create an `SingleCellExperiment` object containing your data. 

<!--
For Tabula muris data we read from files
```{r}

tm.droplet.matrix = readRDS(here("data-tabula-muris", "TM_droplet_mat.rds"))
tm.droplet.metadata = read_csv(here("data-tabula-muris", "TM_droplet_metadata.csv"))

tm.facs.matrix = readRDS(here("data-tabula-muris", "TM_facs_mat.rds"))
tm.facs.metadata = read_csv(here("data-tabula-muris", "TM_facs_metadata.csv"))
# 
# microwell.matrix = readRDS(here("data-microwell-seq", "MCA_merged_mat.rds"))
# microwell.metadata = read_csv(here("data-microwell-seq", "MCA_All-batch-removed-assignments.csv"))

```

The list of tissues available can be found by
```{r}
tiss.name.tm.droplet = unique(tm.droplet.metadata$tissue)
tiss.name.tm.facs = unique(tm.facs.metadata$tissue)

# tiss.name.microwellseq = unique(microwell.metadata$Tissue)
```

We choose a tissue of interest, e.g. bladder, and then use that for the analysis:
```{r}
droplet.tiss.interest = "Bladder"
facs.tiss.interest = "Bladder"

if (droplet.tiss.interest %in% tiss.name.tm.droplet){
  tm.droplet.tiss <- tm.droplet.matrix[,tm.droplet.metadata$tissue==droplet.tiss.interest] 
}

if(facs.tiss.interest %in% tiss.name.tm.facs){
  tm.facs.tiss <- tm.facs.matrix[,tm.facs.metadata$tissue==facs.tiss.interest]
}



#microwell.bladder <- microwell.matrix[,microwell.metadata$Tissue=="Bladder"]

```
-->
Next we create the SingleCellExperiment objects.


```{r tm.droplet.sce}
# tm.droplet.metadata.tiss <- tm.droplet.metadata[tm.droplet.metadata$tissue == droplet.tiss.interest,]
# tm.droplet.tiss.ann <- as.data.frame(tm.droplet.metadata.tiss$cell_ontology_class, row.names = tm.droplet.metadata.tiss$cell)
colnames(tm.droplet.tiss.ann) <- "cell_type1"

sce.tm_droplet <- SingleCellExperiment(assays = list(normcounts = as.matrix(tm.droplet.tiss)), colData = tm.droplet.tiss.ann)
counts(sce.tm_droplet) <- normcounts(sce.tm_droplet)
logcounts(sce.tm_droplet) <- log2(normcounts(sce.tm_droplet) + 1)
# use gene names as feature symbols
rowData(sce.tm_droplet)$feature_symbol <- rownames(sce.tm_droplet)
isSpike(sce.tm_droplet, "ERCC") <- grepl("^ERCC-", rownames(sce.tm_droplet))
# remove features with duplicated names - not need here
sce.tm_droplet <- sce.tm_droplet[!duplicated(rownames(sce.tm_droplet)), ]

sce.tm_droplet
```


```{r tm.facs.sce}
# tm.facs.metadata.tiss <- tm.facs.metadata[tm.facs.metadata$tissue == facs.tiss.interest,]
# tm.facs.tiss.ann <- as.data.frame(tm.facs.metadata.tiss$cell_ontology_class, row.names = tm.facs.metadata.tiss$cell)
colnames(tm.facs.tiss.ann) <- "cell_type1"

sce.tm_facs <- SingleCellExperiment(assays = list(normcounts = as.matrix(tm.facs.tiss)), colData = tm.facs.tiss.ann)
counts(sce.tm_facs) <- normcounts(sce.tm_facs)
logcounts(sce.tm_facs) <- log2(normcounts(sce.tm_facs) + 1)
# use gene names as feature symbols
rowData(sce.tm_facs)$feature_symbol <- rownames(sce.tm_facs)
isSpike(sce.tm_facs, "ERCC") <- grepl("^ERCC-", rownames(sce.tm_facs))
# remove features with duplicated names - not need here
sce.tm_facs <- sce.tm_facs[!duplicated(rownames(sce.tm_facs)), ]

sce.tm_facs
```

### Feature selection

Once we have a SingleCellExperiment object we can run scmap. Firstly, we need to select the most informative features (genes) from our input dataset:

```{r tm.droplet}
sce.tm_droplet <- selectFeatures(sce.tm_droplet, suppress_plot = FALSE)
table(rowData(sce.tm_droplet)$scmap_features)
```

```{r tm.facs}
sce.tm_facs <- selectFeatures(sce.tm_facs, suppress_plot = FALSE)
table(rowData(sce.tm_facs)$scmap_features)
```

We will use 500 features for each, as that is the default (the number can be changed using the ```n_features``` parameter).

## scmap-cluster
### Index
The scmap-cluster index of a reference dataset is created by finding the median gene expression for each cluster. By default *scmap* uses the ```cell_type1``` column of the colData slot in the reference to identify clusters. Other columns can be manually selected by adjusting ```cluster_col``` parameter:

```{r t.droplet.indexCluster}
sce.tm_droplet <- indexCluster(sce.tm_droplet)#,cluster_col = colnames(colData(sce.tm_droplet)))
```

```{r t.facs.indexCluster}
sce.tm_facs <- indexCluster(sce.tm_facs)#,cluster_col = colnames(colData(sce.tm_facs)))
```


The function indexCluster automatically writes the ```scmap_cluster_index``` item of the metadata slot of the reference dataset.

```{r tm.droplet.metadata}
head(metadata(sce.tm_droplet)$scmap_cluster_index)
heatmap(as.matrix(metadata(sce.tm_droplet)$scmap_cluster_index))
```


```{r tm.facs.metadata}
head(metadata(sce.tm_facs)$scmap_cluster_index)
heatmap(as.matrix(metadata(sce.tm_facs)$scmap_cluster_index))
```



### Projection
Once the scmap-cluster index has been generated we can use it to project our dataset to itself (just for illustrative purposes). This can be done with one index at a time, but scmap also allows for simultaneous projection to multiple indexes if they are provided as a list:

```{r tm.droplet.projection}
scmapCluster_results.tm_droplet <- scmapCluster(
  projection = sce.tm_droplet, 
  index_list = list(
    tm_droplet = metadata(sce.tm_droplet)$scmap_cluster_index
  )
)

# scmap-cluster projects the query dataset to all projections defined in the index_list. The results of cell label assignements are merged into one matrix:
head(scmapCluster_results.tm_droplet$scmap_cluster_labs)

# Corresponding similarities are stored in the scmap_cluster_siml item:
head(scmapCluster_results.tm_droplet$scmap_cluster_siml)

# scmap also provides combined results of all reference dataset (choose labels corresponding to the largest similarity across reference datasets):
head(scmapCluster_results.tm_droplet$combined_labs)

# Visualisation
plot(
  getSankey(
    colData(sce.tm_droplet)$cell_type1, 
    scmapCluster_results.tm_droplet$scmap_cluster_labs[,'tm_droplet'],
    plot_height = 400,
    colors = c('#009ACD', '#9ACD32', '#FF8247','#D02090', '#636363')
  )
)
```

```{r tm.facs.projection}
scmapCluster_results.tm_facs <- scmapCluster(
  projection = sce.tm_facs, 
  index_list = list(
    tm_facs = metadata(sce.tm_facs)$scmap_cluster_index
  )
)

# scmap-cluster projects the query dataset to all projections defined in the index_list. The results of cell label assignements are merged into one matrix:
head(scmapCluster_results.tm_facs$scmap_cluster_labs)

# Corresponding similarities are stored in the scmap_cluster_siml item:
head(scmapCluster_results.tm_facs$scmap_cluster_siml)

# scmap also provides combined results of all reference dataset (choose labels corresponding to the largest similarity across reference datasets):
head(scmapCluster_results.tm_facs$combined_labs)

# Visualisation
plot(
  getSankey(
    colData(sce.tm_facs)$cell_type1, 
    scmapCluster_results.tm_facs$scmap_cluster_labs[,'tm_facs'],
    plot_height = 400,
    colors = c('#009ACD', '#9ACD32', '#FF8247','#D02090', '#636363')
  )
)
```



## scmap-cell
In contrast to scmap-cluster, scmap-cell projects cells of the input dataset to the individual cells of the reference and not to the cell clusters.

### Stochasticity
scmap-cell contains k-means step which makes it stochastic, i.e. running it multiple times will provide slightly different results. 

### Index
In the scmap-cell index is created by a product quantiser algorithm in a way that every cell in the reference is identified with a set of sub-centroids found via k-means clustering based on a subset of the features.


```{r tm.droplet.indexCell}
# Fix a random seed to exactly reproduce results:
set.seed(1)
sce.tm_droplet <- indexCell(sce.tm_droplet)
```
```{r tm.facs.indexCell}
# Fix a random seed to exactly reproduce results:
set.seed(1)
sce.tm_facs <- indexCell(sce.tm_facs)
```


Unlike scmap-cluster index, scmap-cell index contains information about each cell and therefore cannot be easily visualised. scmap-cell index consists of two items:

```{r}
names(metadata(sce.tm_droplet)$scmap_cell_index)
names(metadata(sce.tm_facs)$scmap_cell_index)
```


#### Sub-centroids
subcentroids contains coordinates of subcentroids of low dimensional subspaces defined by selected features, k and M parameters of the product quantiser algorithm (see ?indexCell).


```{r tm.droplet.subcentroids}
length(metadata(sce.tm_droplet)$scmap_cell_index$subcentroids)
dim(metadata(sce.tm_droplet)$scmap_cell_index$subcentroids[[1]])
metadata(sce.tm_droplet)$scmap_cell_index$subcentroids[[1]][,1:5]
```

```{r tm.facs.subcentroids}
length(metadata(sce.tm_facs)$scmap_cell_index$subcentroids)
dim(metadata(sce.tm_facs)$scmap_cell_index$subcentroids[[1]])
metadata(sce.tm_facs)$scmap_cell_index$subcentroids[[1]][,1:5]
```


#### Sub-clusters
subclusters contains for every low dimensial subspace indexies of subcentroids which a given cell belongs to:


```{r tm.droplet.subclusters}
dim(metadata(sce.tm_droplet)$scmap_cell_index$subclusters)
metadata(sce.tm_droplet)$scmap_cell_index$subclusters[1:5,1:5]
```

```{r tm.facs.subclusters}
dim(metadata(sce.tm_facs)$scmap_cell_index$subclusters)
metadata(sce.tm_facs)$scmap_cell_index$subclusters[1:5,1:5]
```


### Projection
Once the scmap-cell indexes have been generated we can use them to project the baron dataset. This can be done with one index at a time, but scmap allows for simultaneous projection to multiple indexes if they are provided as a list:


```{r}
scmapCell_results.tm_droplet <- scmapCell(
  sce.tm_droplet, 
  list(
    tm_droplet = metadata(sce.tm_droplet)$scmap_cell_index
  )
)
```

```{r}
scmapCell_results.tm_facs <- scmapCell(
  sce.tm_facs, 
  list(
    tm_facs = metadata(sce.tm_facs)$scmap_cell_index
  )
)
```

### Results
scmapCell_results contains results of projection for each reference dataset in a list:

```{r}
names(scmapCell_results.tm_droplet)
scmapCell_results.tm_droplet$tm_droplet$cells[,1:3]
scmapCell_results.tm_droplet$tm_droplet$similarities[,1:3]
```


```{r}
names(scmapCell_results.tm_facs)
scmapCell_results.tm_facs$tm_facs$cells[,1:3]
scmapCell_results.tm_facs$tm_facs$similarities[,1:3]
```


For each dataset there are two matricies. Cells matrix contains the top 10 (scmap default) cell IDs of the cells of the reference dataset that a given cell of the projection dataset is closest to and Similarities matrix contains corresponding cosine similarities.


### Cluster annotation
If cell cluster annotation is available for the reference datasets, in addition to finding top 10 nearest neighbours scmap-cell also allows to annotate cells of the projection dataset using labels of the reference. It does so by looking at the top 3 nearest neighbours (scmap default) and if they all belong to the same cluster in the reference and their maximum similarity is higher than a threshold (0.5 is the scmap default) a projection cell is assigned to a corresponding reference cluster:

```{r}
scmapCell_clusters.tm_droplet <- scmapCell2Cluster(
  scmapCell_results.tm_droplet, 
  list(
    as.character(colData(sce.tm_droplet)$cell_type1)
  )
)
```

```{r}
scmapCell_clusters.tm_facs <- scmapCell2Cluster(
  scmapCell_results.tm_facs, 
  list(
    as.character(colData(sce.tm_facs)$cell_type1)
  )
)
```


scmap-cell results are in the same format as the ones provided by scmap-cluster (see above):

```{r}
head(scmapCell_clusters.tm_droplet$scmap_cluster_labs)
```
```{r}
head(scmapCell_clusters.tm_facs$scmap_cluster_labs)
```

Corresponding similarities are stored in the scmap_cluster_siml item:


```{r}
head(scmapCell_clusters.tm_droplet$scmap_cluster_siml)
head(scmapCell_clusters.tm_droplet$combined_labs)
```

```{r}
head(scmapCell_clusters.tm_facs$scmap_cluster_siml)
head(scmapCell_clusters.tm_facs$combined_labs)
```


### Visualisation

```{r}
plot(
  getSankey(
    colData(sce.tm_droplet)$cell_type1, 
    scmapCell_clusters.tm_droplet$scmap_cluster_labs[,"tm_droplet"],
    plot_height = 400,
    colors = c('#009ACD', '#9ACD32', '#FF8247','#D02090', '#636363')
  )
)
```

```{r}
plot(
  getSankey(
    colData(sce.tm_facs)$cell_type1, 
    scmapCell_clusters.tm_facs$scmap_cluster_labs[,"tm_facs"],
    plot_height = 400,
    colors = c('#009ACD', '#9ACD32', '#FF8247','#D02090', '#636363')
    # colors = rev(c('#ffffd9','#edf8b1','#c7e9b4','#7fcdbb','#41b6c4','#1d91c0','#225ea8','#253494','#081d58'))
  )
)

```










## SessionInfo
```{r}
sessionInfo()
```

